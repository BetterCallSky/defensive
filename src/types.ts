import { V, Schema, Convert } from 'veni';
import * as Logger from 'bunyan';
import { ContractBinding } from './ContractBinding';

interface Contract2<
  ARG1 extends string,
  ARG2 extends string,
  TSchema extends { [key in ARG1 | ARG2]: Schema }
> {
  schema<T extends { [key in ARG1 | ARG2]: Schema }>(
    param: T
  ): Contract2<ARG1, ARG2, T>;

  fn<
    T extends (arg1: Convert<TSchema[ARG1]>, arg2: Convert<TSchema[ARG2]>) => R,
    R
  >(
    fn: T
  ): T;
}

export interface ContractMeta<TSchema> {
  getSchema(): TSchema;
}

interface Contract1<
  ARG1 extends string,
  TSchema extends { [key in ARG1]: Schema }
> {
  schema<T extends { [key in ARG1]: Schema }>(param: T): Contract1<ARG1, T>;

  fn<T extends (arg1: Convert<TSchema[ARG1]>) => R, R>(
    fn: T
  ): T & ContractBinding<T> & ContractMeta<TSchema>;
}

export interface Contract {
  config(config: Partial<ContractConfig>): this;
  options(options: Partial<ContractOptions>): this;
  params<ARG1 extends string>(arg1: ARG1): Contract1<ARG1, any>;
  params<ARG1 extends string, ARG2 extends string>(
    arg1: ARG1,
    arg2: ARG2
  ): Contract2<ARG1, ARG2, any>;
}

export interface ContractOptions {
  sync: boolean;
  removeOutput: boolean;
}

export interface ContractConfig {
  removeFields: string[];
  debug: boolean;
  depth: number;
  maxArrayLength: number;
  getLogger: (serviceName: string) => Logger;
  getNextId: () => number;
}
