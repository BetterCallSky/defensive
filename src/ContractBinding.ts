export class ContractBinding<T> {
  fn?: T;
}
