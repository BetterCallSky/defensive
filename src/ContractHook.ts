import async_hooks from 'async_hooks';

export class ContractHook {
  private asyncHook: async_hooks.AsyncHook = null!;
  private rootMapping: Map<number, number> = new Map();
  private groups: Map<number, number[]> = new Map();

  constructor() {
    this.asyncHook = async_hooks.createHook({
      init: (asyncId, type, triggerAsyncId, resource) => {
        if (!this.rootMapping.has(triggerAsyncId)) {
          return;
        }
        const rootId = this.rootMapping.get(triggerAsyncId)!;
        this.rootMapping.set(asyncId, rootId);
        if (!this.groups.has(rootId)) {
          this.groups.set(rootId, []);
        }
        this.groups.get(rootId)!.push(asyncId);
      },
      destroy: asyncId => {
        if (!this.rootMapping.has(asyncId)) {
          return;
        }
        const rootId = this.rootMapping.get(asyncId)!;
        if (rootId !== asyncId) {
          return;
        }
        const ids = this.groups.get(asyncId) || [];
        ids.forEach(id => {
          this.rootMapping.delete(id);
        });
        this.groups.delete(asyncId);
      },
    });
  }

  addRoot(asyncId: number) {
    this.rootMapping.set(asyncId, asyncId);
  }

  enable() {
    this.asyncHook.enable();
  }

  disable() {
    this.asyncHook.disable();
  }
}
