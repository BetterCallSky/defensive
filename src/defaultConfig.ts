// import * as Logger from 'bunyan';
// import { ContractConfig } from './types';

// let defaultId = 0;

// const loggerMap: Record<string, Logger> = {};

// const defaultConfig: ContractConfig = {
//   removeFields: ['password', 'token', 'accessToken'],
//   debug: true,
//   depth: 4,
//   maxArrayLength: 30,
//   getLogger(serviceName: string) {
//     if (!loggerMap[serviceName]) {
//       loggerMap[serviceName] = Logger.createLogger({
//         name: serviceName,
//         level: this.debug ? 'debug' : 'error',
//       });
//     }
//     return loggerMap[serviceName];
//   },
//   getNextId: () => ++defaultId,
// };

// export const setDefaultConfig = (config: Partial<ContractConfig>) => {
//   Object.assign(defaultConfig, config);
// };

// export const getDefaultConfig = () => defaultConfig;
