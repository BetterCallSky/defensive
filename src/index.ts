export * from './createContract';
export * from './defaultConfig';
export * from './types';
export * from './ContractBinding';
