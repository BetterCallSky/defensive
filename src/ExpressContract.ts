import { ContractBinding } from './ContractBinding';
import { Router, Request, Response } from 'express';

ContractBinding.prototype.express = function(options) {
  if (!this.fn.expressOptions) {
    this.fn.expressOptions = [];
  }
  this.fn.expressOptions.push(options);
  return this.fn as any;
};

interface TestThis<T> {
  fn: T;
}

interface ExpressOptions {
  method: 'get' | 'post' | 'put' | 'delete' | 'path';
  path: string;
  handler(req: Request, res: Response): void;
}

declare module './ContractBinding' {
  interface ContractBinding<T> {
    expressOptions: ExpressOptions[];
    express(options: ExpressOptions): T & ContractBinding<T>;
  }
}
