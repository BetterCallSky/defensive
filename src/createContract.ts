import {
  Contract,
  ContractMeta,
  ContractConfig,
  ContractOptions,
} from './types';
import { wrapValidate } from './wrapValidate';
import { wrapLog } from './wrapLog';
import { SchemaMap } from 'veni';
import { ContractBinding } from './ContractBinding';
import { getDefaultConfig } from './defaultConfig';

export const createContract = (signature: string) => {
  const [serviceName, methodName] = signature.split('#');
  if (!methodName) {
    throw new Error(
      `Invalid signature. Must be in format "service-name#method-name". Received "${signature}".`
    );
  }
  const contract = {} as any;
  let config: ContractConfig = getDefaultConfig();
  let options: ContractOptions = {
    sync: false,
    removeOutput: false,
  };
  let params: string[] = [];
  let schema: SchemaMap | null = null;

  const meta: ContractMeta<any> = {
    getSchema: () => schema,
  };

  contract.params = (...args: string[]) => {
    params = args;
    return contract;
  };

  contract.schema = (arg: any) => {
    schema = arg;
    return contract;
  };

  contract.config = (value: any) => {
    config = { ...config, ...value };
    return contract;
  };

  contract.options = (value: any) => {
    options = { ...options, ...value };
    return contract;
  };

  contract.fn = (fn: any) => {
    const wrappedFunction = (...args: any[]) => {
      const logger = config.getLogger(serviceName);
      const withValidation = wrapValidate({
        keysSchema: schema,
        method: fn,
        paramNames: params,
        sync: options.sync,
      });

      const withLogging = wrapLog({
        logger,
        method: withValidation,
        methodName: methodName,
        paramNames: params,
        config,
        removeOutput: options.removeOutput,
      });
      return withLogging(...args);
    };
    const contractBinding = new ContractBinding();
    contractBinding.fn = wrappedFunction;
    const functions = Object.keys(ContractBinding.prototype);
    const ret = Object.assign(wrappedFunction, meta) as any;

    functions.forEach(name => {
      ret[name] = (contractBinding as any)[name].bind(contractBinding);
    });
    return ret;
  };

  return contract as Contract;
};
