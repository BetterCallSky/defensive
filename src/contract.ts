import './ExpressContract';
import { createContract } from './createContract';
import { V } from 'veni';

export const inc = createContract('inc')
  .params('n')
  .schema({
    n: V.number(),
  })
  .fn(n => {
    return n + 1;
  })
  .express({
    method: 'get',
    path: '/inc',
    handler: (req, res) =>
      res.json({
        result: inc(req.body.id),
      }),
  });

export const registerUser = createContract('registerUser')
  .params('values')
  .schema({
    values: V.object().keys({
      username: V.string()
        .min(3)
        .max(10),
      password: V.string()
        .min(5)
        .optional(),
    }),
  })
  .fn(async values => {
    return {
      id: 10,
      username: values.username,
    };
  })
  .express({
    method: 'post',
    path: '/register',
    handler: async (req, res) =>
      res.json({
        user: await registerUser(req.body),
      }),
  });

async function start() {
  await registerUser({
    username: 'foo',
    password: 'aabaaba',
  });
}

start();
