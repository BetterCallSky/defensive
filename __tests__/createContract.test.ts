import * as Logger from 'bunyan';
import { V } from 'veni';
import { Contract } from '../src/types';
import { createContract } from '../src/createContract';
import { setDefaultConfig } from '../src/defaultConfig';

let logger: Logger;

beforeEach(() => {
  logger = getLogger();
  let defaultId = 0;
  setDefaultConfig({
    getLogger: () => logger,
    getNextId: () => ++defaultId,
  });
});

function getLogger() {
  return ({
    error: jest.fn(),
    debug: jest.fn(),
  } as any) as Logger;
}

it('should throw if signature is invalid', () => {
  expect(() => {
    createContract('foo')
      .options({ sync: true })
      .params()
      .fn(() => 10);
  }).toThrowErrorMatchingInlineSnapshot(
    `"Invalid signature. Must be in format \\"service-name#method-name\\". Received \\"foo\\"."`
  );
});

it('validate sync function', () => {
  const fn = createContract('myService#fn')
    .options({ sync: true })
    .params('a')
    .schema({
      a: V.number().min(0),
    })
    .fn(a => a + 10);
  expect(() => {
    fn(10);
  }).not.toThrow();
  expect(() => {
    fn(-10);
  }).toThrowErrorMatchingInlineSnapshot(
    `"Validation error: 'a' must be greater or equal to 0."`
  );
  expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
  expect((logger.error as jest.Mock<any>).mock.calls).toMatchSnapshot();
});

it('validate sync with nested object', () => {
  const fn = createContract('myService#fn')
    .options({ sync: true })
    .params('params')
    .schema({
      params: V.object().keys({
        username: V.string().min(2),
        inner: V.object().keys({
          b: V.number().min(0),
        }),
      }),
    })
    .fn(params => params.username + '-' + params.inner.b);
  expect(() => {
    fn({ username: 'baz', inner: { b: -1 } });
  }).toThrowErrorMatchingInlineSnapshot(
    `"Validation error: 'params.inner.b' must be greater or equal to 0."`
  );
  expect(() => {
    fn({ username: 'b', inner: { b: 10 } });
  }).toThrowErrorMatchingInlineSnapshot(
    `"Validation error: 'params.username' length must be at least 2 characters long."`
  );
  expect(fn({ username: 'bar', inner: { b: 10 } })).toEqual('bar-10');
});

it('validate async function', async () => {
  const fn = createContract('myService#fn')
    .params('a')
    .schema({
      a: V.number().min(0),
    })
    .fn(async a => a + 10);
  await expect(fn(10)).resolves.toBe(20);
  await expect(fn(-10)).rejects.toThrowErrorMatchingInlineSnapshot(
    `"Validation error: 'a' must be greater or equal to 0."`
  );
  expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
  expect((logger.error as jest.Mock<any>).mock.calls).toMatchSnapshot();
});

it('with removeOutput', () => {
  const fn = createContract('myService#fn')
    .options({ sync: true, removeOutput: true })
    .params('a')
    .schema({
      a: V.number().min(0),
    })
    .fn(a => a + 10);

  fn(10);
  expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
});
