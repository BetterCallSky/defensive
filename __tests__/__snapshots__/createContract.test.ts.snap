// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate async function 2`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER fn:",
    "{ a: 10 }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT fn:",
    "20",
  ],
  Array [
    Object {
      "id": 2,
    },
    "ENTER fn:",
    "{ a: -10 }",
  ],
]
`;

exports[`validate async function 3`] = `
Array [
  Array [
    Object {
      "id": 2,
    },
    "ERROR fn: { a: -10 } 
",
    [Error: Validation error: 'a' must be greater or equal to 0.],
  ],
]
`;

exports[`validate sync function 2`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER fn:",
    "{ a: 10 }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT fn:",
    "20",
  ],
  Array [
    Object {
      "id": 2,
    },
    "ENTER fn:",
    "{ a: -10 }",
  ],
]
`;

exports[`validate sync function 3`] = `
Array [
  Array [
    Object {
      "id": 2,
    },
    "ERROR fn: { a: -10 } 
",
    [Error: Validation error: 'a' must be greater or equal to 0.],
  ],
]
`;

exports[`with removeOutput 1`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER fn:",
    "{ a: 10 }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT fn:",
    "<removed>",
  ],
]
`;
